torcheeg/trainers/classifier.py

```diff
@@ -266,4 +296,349 @@ def predict_step(self,
         x, y = batch
         y_hat = self(x)
-        return y_hat
\ No newline at end of file
+        return y_hat
+
+
+class CLossClassifierTrainer(ClassifierTrainer):
+    r'''
+        A trainer trains classification model whose loss function contains Center loss function. PLease refer to the following infomation to comprehend how the center loss works.
+
+        - Paper: FBMSNet: A Filter-Bank Multi-Scale Convolutional Neural Network for EEG-Based Motor Imagery Decoding
+        - URL: https://ieeexplore.ieee.org/document/9837422
+        - Related Project: https://github.com/Want2Vanish/FBMSNet
+
+        
+        The model structure is required to contains a decoder block which generates the deep feature code and a predictor connected to the decoder to judge which class the feature code belong to.
+        Firstly, we should prepare a model which contains :obj:`decoder` , :obj:`"predict_by_feature"` attributes whose value belongs to :obj:`nn.Module` subclass.
+
+        .. code-block:: python
+
+            class MyModel(nn.Module):
+                def __init__(self) -> None:
+                    super(md,self).__init__()
+                    self.decoder = nn.Linear(10,5)
+                    self.predict_by_feature= nn.Linear(5,2)
+    
+                def forward(self,x):
+                    return self.predict_by_feature(self.decoder(x))
+            model = MyModel()
+        
+        Use :obj:`nn.ModuleDict` is a simplest way to define model.
+
+        .. code-block:: python
+
+            model = nn.ModuleDict( {'decoder': nn.Linear(10,5),'predict_by_feature': nn.Linear(5,2)})
+
+                    
+        And it is also feasible to define :obj:`decorder` and :obj:`predict_by_feature` method in your model.  
+
+        .. code-block:: python
+
+            class MyModel(nn.Module):
+                def __init__(self) -> None:
+                    super(md,self).__init__()
+                    self.layer1 = nn.Linear(10,5)
+                    self.layer2 = nn.Linear(5,2)
+                
+                def decoder(self,x):
+                    return self.layer1(x)
+                
+                def predict_by_feature(feature):
+                    return self.layer2(feature)
+    
+                def forward(self,x):
+                    return self.predict_by_feature(self.decoder(x))
+            model = MyModel()
+        
+
+        Then we simply pass the model to init trainer. Done!
+        
+        .. code-block:: python
+
+            trainer = ClossClassifierTrainer(model)
+            trainer.fit(train_loader, val_loader)
+            trainer.test(test_loader)
+
+        Args:
+            model (nn.Module): model (nn.Module): The classification model, which should provide the :obj:`".decoder()"` , :obj:`".predict_by_feature()"` method or the model  should contain decoder block(nn.Module), predict block(nn.Module) attributes and ensure this two attirbutes are named "decoder","predict_by_feature".
```

Comment:
The requirements seem overly stringent. Please refer to the design of the domain adaptation trainer. You may ask the user to input two modules.
```
class ADATrainer(_MMDLikeTrainer):
    r'''
    This class supports the implementation of Associative Domain Adaptation (ADA) for deep domain adaptation.

    NOTE: ADA belongs to unsupervised domain adaptation methods, which only use labeled source data and unlabeled target data. This means that the target dataset does not have to contain labels.

    - Paper: Haeusser P, Frerix T, Mordvintsev A, et al. Associative domain adaptation[C]//Proceedings of the IEEE international conference on computer vision. 2017: 2765-2773.
    - URL: https://arxiv.org/abs/1708.00938
    - Related Project: https://github.com/stes/torch-assoc

    .. code-block:: python

        trainer = ADATrainer(extractor,
                            classifier,
                            num_classes=10,
                            devices=1,
                            weight_visit=0.6,
                            accelerator='gpu')
        trainer.fit(source_loader, target_loader, val_loader)
        trainer.test(test_loader)

    Args:
        extractor (nn.Module): The feature extraction model learns the feature representation of the EEG signal by forcing the correlation matrixes of source and target data to be close.
        classifier (nn.Module): The classification model learns the classification task with the source labeled data based on the feature of the feature extraction model. The dimension of its output should be equal to the number of categories in the dataset. The output layer does not need to have a softmax activation function.
        num_classes (int, optional): The number of categories in the dataset. (default: :obj:`None`)
        lr (float): The learning rate. (default: :obj:`0.0001`)
        weight_decay (float): The weight decay. (default: :obj:`0.0`)
        weight_walker (float): The weight of the walker loss. (default: :obj:`1.0`)
        weight_visit (float): The weight of the visit loss. (default: :obj:`1.0`)
        weight_domain (float): The weight of the associative loss (default: :obj:`1.0`)
        weight_scheduler (bool): Whether to use a scheduler for the weight of the associative loss, growing from 0 to 1 following the schedule from the DANN paper. (default: :obj:`False`)
        lr_scheduler (bool): Whether to use a scheduler for the learning rate, as defined in the DANN paper. (default: :obj:`False`)
        warmup_epochs (int): The number of epochs for the warmup phase, during which the weight of the associative loss is 0. (default: :obj:`0`)
        devices (int): The number of devices to use. (default: :obj:`1`)
        accelerator (str): The accelerator to use. Available options are: 'cpu', 'gpu'. (default: :obj:`"cpu"`)
        metrics (list of str): The metrics to use. Available options are: 'precision', 'recall', 'f1score', 'accuracy'. (default: :obj:`["accuracy"]`)

    .. automethod:: fit
    .. automethod:: test
    '''
    def __init__(self,
                 extractor: nn.Module, # see here
                 classifier: nn.Module,  # see here
                 num_classes: int,
                 lr: float = 1e-4,
                 weight_walker: float = 1.0,
                 weight_visit: float = 1.0,
                 weight_domain: float = 1.0,
                 weight_decay: float = 0.0,
                 weight_scheduler: bool = True,
                 lr_scheduler_gamma: float = 0.0,
                 lr_scheduler_decay: float = 0.75,
                 warmup_epochs: int = 0,
                 devices: int = 1,
                 accelerator: str = "cpu",
                 metrics: List[str] = ["accuracy"]):

        super(ADATrainer,
              self).__init__(extractor=extractor,
                             classifier=classifier,
                             num_classes=num_classes,
                             lr=lr,
                             weight_decay=weight_decay,
                             weight_domain=weight_domain,
                             weight_scheduler=weight_scheduler,
                             lr_scheduler_gamma=lr_scheduler_gamma,
                             lr_scheduler_decay=lr_scheduler_decay,
                             warmup_epochs=warmup_epochs,
                             devices=devices,
                             accelerator=accelerator,
                             metrics=metrics)
        self.weight_walker = weight_walker
        self.weight_visit = weight_visit

        self._assoc_fn = AssociativeLoss(weight_walker, weight_visit)
```

Reply:
Some models define the decoder and classifier in single class, like FBMS , is it ok to make the input type of decoder and classifier includes method or nn.modules ?

Reply:
Please refer to this [document](https://torcheeg.readthedocs.io/en/latest/auto_examples/examples_seed_domain_adaption.html#sphx-glr-auto-examples-examples-seed-domain-adaption-py):

```
from torcheeg.models import [CCNN](https://torcheeg.readthedocs.io/en/latest/generated/torcheeg.models.CCNN.html#torcheeg.models.CCNN)


class Extractor([CCNN](https://torcheeg.readthedocs.io/en/latest/generated/torcheeg.models.CCNN.html#torcheeg.models.CCNN)):
    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = self.conv3(x)
        x = self.conv4(x)
        x = x.flatten(start_dim=1)
        return x


class Classifier([CCNN](https://torcheeg.readthedocs.io/en/latest/generated/torcheeg.models.CCNN.html#torcheeg.models.CCNN)):
    def forward(self, x):
        x = self.lin1(x)
        x = self.lin2(x)
        return x

extractor = Extractor(in_channels=5, num_classes=3)
classifier = Classifier(in_channels=5, num_classes=3)

trainer = [ADATrainer](https://torcheeg.readthedocs.io/en/latest/generated/torcheeg.trainers.ADATrainer.html#torcheeg.trainers.ADATrainer)(extractor=extractor,
                             classifier=classifier,
                             num_classes=2,
                             lr=1e-4,
                             weight_decay=0.0,
                             accelerator='gpu')
```